ggml.c

```diff
@@ -1123,12 +1124,94 @@ static void quantize_row_q4_2_reference(const float * restrict x, block_q4_2 * r
     }
 }
 
+static inline int nearestInt(float fval) {
```

Comment:
So, I asked @ggerganov if his preference is to add a new quantization type (e.g., `Q4_2rmse` or similar) or to just overwrite the existing reference implementation. He told me to do the latter. Now `Q4_2` is no longer the same as the original `Q4_2`, so there is no need to try to match the `roundf` behavior as what comes out is no longer just the result of `roundf(7/amax)`.

Reply:
Concerning optimizing RMSE (or not): yes, I agree that it is not clear if RMSE minimization really is the best strategy. But we do get an improvement in perplexity that is kind of consistent with the reduction in quantization RMSE, so that seems to indicate that the strategy is better than nothing. One can try minimizing MAD (mean absolute difference), max. difference between actual and quantized weights, difference between some moment of the actual and quantized weights, etc. One can also, as you suggest, just try running perplexity using different scales. But that would be exceedingly expensive even on a fast computer, without it being clear if it will lead to success (RMSE minimization shows that the optimum scale can be very different between different quantization "buckets", so at least to me, it is not really clear if a global scale different from `7` applied to all model weights when quantizing will lead to improvements).

Reply:
Let's leave it as it is

When we start finalizing the `Q4_2` and `Q4_3` implementations will decide if `nearest_int()` speed is really necessary or we switch to `roundf()` for simplicity.

@sw 
The RMSE optimization brings about `-0.03` ppl delta which is difficult to ignore.
Now that we do the quantization only offline (i.e. when generating the model), we can afford more complex quantization schemes like this one. Closing the gap between quantized and FP16 perplexity is important